go/k8s/controllers/user_controller.go

```diff
@@ -100,66 +142,144 @@ func (r *UserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.
+		userStatus       backendv1.UserStatus
+		retrievedUser    *backendv1.User
 	)
 
 	defer func() {
 		if err != nil {
 			userStatus.State = err.Error()
 		}
 		if retrievedUser != nil {
+			if r.almondConfig.EnableDeveloperBackend && trustedDeveloper {
+				retrievedUser.Spec.Mode = "developer"
+			} else {
+				retrievedUser.Spec.Mode = "shared"
+			}
+			r.Update(ctx, retrievedUser)
 			retrievedUser.Status = userStatus
 			r.Status().Update(ctx, retrievedUser)
 			r.Log.Info("update status:", "user", retrievedUser.Spec.ID, "status", retrievedUser.Status)
 		}
 		r.Log.Info("--- end ---")
 	}()
 
-	backendURL, userID, err = r.getBackendURL(ctx, req.NamespacedName.Name, req)
+	userID, trustedDeveloper, err = r.getUserFromName(req.Name)
 	if err != nil {
-		return ctrl.Result{RequeueAfter: 2 * time.Second}, err
+		return ctrl.Result{}, err
 	}
+
+	if r.almondConfig.EnableDeveloperBackend && trustedDeveloper {
+		// for developer users, make sure deployment and service are up before proceeding.
+		user := &backendv1.User{}
+		if err = r.Client.Get(ctx, req.NamespacedName, user); err != nil {
+			if apierrors.IsNotFound(err) {
+				if err = r.deleteDeploymentService(ctx, req, userID); err != nil {
+					r.Log.Error(err, "fail to delete developer deployment or service")
+				}
+			}
+			return ctrl.Result{}, err
+		}
+		retrievedUser = user
+		if !retrievedUser.ObjectMeta.DeletionTimestamp.IsZero() {
+			// user is marked for deletion
+			user.Status.State = string(Stopping)
+			if err = r.deleteDeploymentService(ctx, req, userID); err != nil {
+				r.Log.Error(err, "fail to delete developer deployment or service marked for deletion")
+			}
+			return ctrl.Result{}, err
+		}
+		deployment := &appsv1.Deployment{}
+		if err = r.Client.Get(ctx, req.NamespacedName, deployment); err != nil {
+			if apierrors.IsNotFound(err) {
+				if err = r.createDeployment(ctx, req.Name, req.Namespace); err != nil {
+					return ctrl.Result{}, err
+				}
+				return ctrl.Result{RequeueAfter: 1 * time.Second}, nil
+			}
+			return ctrl.Result{}, err
+		}
+		if deployment.Status.AvailableReplicas <= 0 {
+			err = errors.New("deployment not ready yet")
+			return ctrl.Result{RequeueAfter: 2 * time.Second}, nil
+		}
+		service := &corev1.Service{}
+		if err = r.Client.Get(ctx, req.NamespacedName, service); err != nil {
+			if apierrors.IsNotFound(err) {
+				if err = r.createService(ctx, req.Name, req.Namespace); err != nil {
+					return ctrl.Result{}, err
+				}
+				err = errors.New("service not ready")
+				return ctrl.Result{RequeueAfter: 1 * time.Second}, nil
+			}
+			return ctrl.Result{}, err
+		}
+		if len(service.Spec.ClusterIP) == 0 {
+			err = errors.New("service ip not ready")
+			return ctrl.Result{RequeueAfter: 2 * time.Second}, nil
+		}
+		if len(service.Spec.Ports) == 0 {
+			err = errors.New("service port not available")
+			return ctrl.Result{}, err
+		}
+		backendURL = fmt.Sprintf("http://%s:%d", service.Spec.ClusterIP, service.Spec.Ports[0].Port)
```

Comment:
The service has a predictable name we set, so why do we need to fetch the cluster IP? coredns will resolve the service name for us.

Reply:
One thing I'm wondering about, what if the pod gets killed for reasons unrelated to the controller (say, the cluster-autoscaler decides to migrate pods to take the node down, or the node dies for random reasons)? Are we watching the endpoints of the service to update the IP address?

Reply:
Yes, it’s updated on every reconcile. Every 10 seconds when running.